File: stroom-query-api/src/main/java/stroom/query/api/DocRef.java
/*
 * Copyright 2016 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package stroom.query.api;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;
import stroom.util.shared.HasDisplayValue;

import javax.xml.bind.annotation.XmlRootElement;
import javax.xml.bind.annotation.XmlTransient;
import javax.xml.bind.annotation.XmlType;
import java.io.Serializable;

@JsonPropertyOrder({"type", "id", "uuid", "name"})
@XmlType(name = "DocRef", propOrder = {"type", "id", "uuid", "name"})
@XmlRootElement(name = "doc")
public class DocRef implements Comparable<DocRef>, HasDisplayValue, Serializable {
    private static final long serialVersionUID = -2121399789820829359L;

    protected String type;
    protected String uuid;
    protected String name;

    @Deprecated
    protected Long id;

    public DocRef() {
    }

    public DocRef(final String type, String uuid) {
        this.type = type;
        this.uuid = uuid;
    }

    public DocRef(final String type, String uuid, final String name) {
        this.type = type;
        this.uuid = uuid;
        this.name = name;
    }

    public String getType() {
        return type;
    }

    public void setType(final String value) {
        this.type = value;
    }

    public String getUuid() {
        return uuid;
    }

    public void setUuid(final String uuid) {
        this.uuid = uuid;
    }

    public String getName() {
        return name;
    }

    public void setName(final String name) {
        this.name = name;
    }

    @Deprecated
    public Long getId() {
        return id;
    }

    @Deprecated
    public void setId(final Long value) {
        this.id = value;

        // All equality is done on uuid so ensure uuid is set to id even if the entity doesn't have a uuid field.
        if (uuid == null && value != null) {
            uuid = String.valueOf(value);
        }
    }

    @XmlTransient
    @JsonIgnore
    @Override
    public String getDisplayValue() {
        return name;
    }

    @Override
    public int compareTo(final DocRef o) {
        int diff = type.compareTo(o.type);

        if (diff == 0) {
            if (name != null && o.name != null) {
                diff = name.compareTo(o.name);
            }
        }
        if (diff == 0) {
            diff = uuid.compareTo(o.uuid);
        }
        return diff;
    }

    public String toInfoString() {
        final StringBuilder sb = new StringBuilder();
        if (name != null) {
            sb.append(name);
        }
        if (uuid != null) {
            if (sb.length() > 0) {
                sb.append(" ");
            }
            sb.append("{");
            sb.append(uuid);
            sb.append("}");
        }

        if (sb.length() > 0) {
            return sb.toString();
        }

        return toString();
    }


    @Override
    public boolean equals(final Object o) {
        if (this == o) return true;
        if (!(o instanceof DocRef)) return false;

        final DocRef docRef = (DocRef) o;

        if (type != null ? !type.equals(docRef.type) : docRef.type != null) return false;
        if (uuid != null ? !uuid.equals(docRef.uuid) : docRef.uuid != null) return false;
        return name != null ? name.equals(docRef.name) : docRef.name == null;
    }

    @Override
    public int hashCode() {
        int result = type != null ? type.hashCode() : 0;
        result = 31 * result + (uuid != null ? uuid.hashCode() : 0);
        result = 31 * result + (name != null ? name.hashCode() : 0);
        return result;
    }

    @Override
    public String toString() {
        return "DocRef{" +
                "type='" + type + '\'' +
                ", uuid='" + uuid + '\'' +
                ", name='" + name + '\'' +
                '}';
    }
}

File: stroom-ui/src/components/DocExplorer/DocRef.js
/*
 * Copyright 2018 Crown Copyright
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import PropTypes from 'prop-types';

import { compose } from 'recompose';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { Icon } from 'semantic-ui-react';

import ItemTypes from './dragDropTypes';
import { DragSource } from 'react-dnd';

import { actionCreators } from './redux';

import DocRefMenu from './DocRefMenu';
import ClickCounter from 'lib/ClickCounter';
import { openDocRef } from 'prototypes/RecentItems';

const {
  docRefSelected,
  docRefContextMenuOpened,
  docRefContextMenuClosed,
} = actionCreators;

const dragSource = {
  canDrag(props) {
    return true;
  },
  beginDrag(props) {
    return {
      docRef: props.docRef,
      isCopy: !!(props.keyIsDown.Control || props.keyIsDown.Meta),
    };
  },
};

function dragCollect(connect, monitor) {
  return {
    connectDragSource: connect.dragSource(),
    isDragging: monitor.isDragging(),
  };
}

const enhance = compose(
  withRouter,
  connect(
    (
      {
        keyIsDown,
        docExplorer: {
          explorerTree: { explorers },
        },
      },
      { explorerId },
    ) => ({
      explorer: explorers[explorerId],
      keyIsDown,
    }),
    {
      docRefSelected,
      openDocRef,
      docRefContextMenuOpened,
      docRefContextMenuClosed,
    },
  ),
  DragSource(ItemTypes.DOC_REF, dragSource, dragCollect),
);

const DocRef = ({
  explorerId,
  explorer,
  docRef,
  history,
  docRefSelected,
  openDocRef,
  docRefContextMenuOpened,
  docRefContextMenuClosed,
  connectDragSource,
  isDragging,
}) => {
  const isSelected = explorer.isSelected[docRef.uuid];
  const { contentMenuDocRef } = explorer;
  const isContextMenuOpen = !!contentMenuDocRef && contentMenuDocRef.uuid === docRef.uuid;
  const isPartOfContextMenuSelection = !!contentMenuDocRef && isSelected;

  // these are required to tell the difference between single/double clicks
  const clickCounter = new ClickCounter()
    .withOnSingleClick(({ appendSelection, contiguousSelection }) =>
      docRefSelected(explorerId, docRef, appendSelection, contiguousSelection))
    .withOnDoubleClick(() => openDocRef(history, docRef));

  const onRightClick = (e) => {
    if (!isSelected) {
      docRefSelected(explorerId, docRef, false, false);
    }
    docRefContextMenuOpened(explorerId, docRef);
    e.preventDefault();
  };

  let className = '';
  if (isDragging) {
    className += ' doc-ref__dragging';
  }
  if (isSelected) {
    className += ' doc-ref__selected';
  }
  if (isPartOfContextMenuSelection) {
    className += ' doc-ref__context-menu-open';
  }

  return connectDragSource(<div
    className={className}
    onContextMenu={onRightClick}
    onDoubleClick={() => clickCounter.onDoubleClick()}
    onClick={e =>
        clickCounter.onSingleClick({
          appendSelection: e.ctrlKey || e.metaKey,
          contiguousSelection: e.shiftKey,
        })
      }
  >
    <DocRefMenu
      explorerId={explorerId}
      docRef={docRef}
      isOpen={isContextMenuOpen}
      closeContextMenu={() => docRefContextMenuClosed(explorerId)}
    />
    <span>
      <Icon />
      <img className="doc-ref__icon" alt="X" src={require(`./images/${docRef.type}.svg`)} />
      {docRef.name}
    </span>
                           </div>);
};

DocRef.propTypes = {
  explorerId: PropTypes.string.isRequired,
  docRef: PropTypes.object.isRequired,
};

export default enhance(DocRef);
